### PR TITLE
feat(query): add timeout option

### LINKED ISSUE #185: Timeouts are not supported
Incompatibility with **felixge** node-mysql, see documentation:
https://github.com/felixge/node-mysql#timeouts

### PR DESCRIPTION
**Pull request by yuxizhe**
add query timeout option. 

> https://github.com/mysqljs/mysql#timeouts

fix #185 #1104 #1349 and so on.

### CODE REVIEW

**Comment by yuxizhe** [comment]
@sidorares Please take a look. 😄

**Comment by sidorares** [comment]
@yuxizhe thanks for working on this! Do you mind adding a unit test?

**Comment by yuxizhe** [comment]
> @yuxizhe thanks for working on this! Do you mind adding a unit test?

no problem,  I am working on it.

**Comment by sidorares** [comment]
@yuxizhe would same timeout parameter work foe `.execute()` as well? Also `.execute()` might be 2 commands under the hood, if we want it to support timeout we need to account for that

**Comment by sidorares** [comment]
what is behaviour when we have handshake timeout? Does `timeout` option allows us to have connection error if handshake is not complete within specified time?

**Comment by yuxizhe** [comment]
> what is behaviour when we have handshake timeout? Does `timeout` option allows us to have connection error if handshake is not complete within specified time?

do you mean  `connectTimeout` and `acquireTimeout` options?  I will add a unit test for `connectTimeout` . But `acquireTimeout` is not support now in `mysql2`.

File: test/integration/connection/test-connect-timeout.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (yuxizhe) [reply test/integration/connection/test-connect-timeout.js:1]: this test the `connectTimeout` option

File: test/integration/connection/test-query-timeout.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+});
+
+const connectionTimeout = common.createConnection({ 
+  host: '10.255.255.1',
+  debug: false,
+  connectTimeout: 100, 
+});
+
+// return connect timeout error first
+connectionTimeout.query({ sql: 'SELECT sleep(3) as a', timeout: 50 }, (err, res) => {
```
[the remark this answers is not in the document]
Comment (yuxizhe) [reply test/integration/connection/test-query-timeout.js:45]: According to sidorares's suggestion, I add a `connectTimeout` test file and put this test here to explain the `timeout` option wont change the connect timeout behavior.

> what is behaviour when we have handshake timeout? Does `timeout` option allows us to have connection error if handshake is not complete within specified time?

File: test/integration/connection/test-connect-timeout.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+const common = require('../../common');
+const connection = common.createConnection({ 
+  host: '10.255.255.1',
```
[the remark this answers is not in the document]
Comment (yuxizhe) [reply test/integration/connection/test-connect-timeout.js:5]: its the same host in `mysqljs`.  PR: `tests: use non-routable host for connect timeout test`

>   https://github.com/mysqljs/mysql/blob/master/test/unit/connection/test-connect-timeout.js#L15
```
var connection = common.createConnection({host: '10.255.255.1', port: common.fakeServerPort, connectTimeout: 500});
```

[the remark this answers is not in the document]
Comment (yuxizhe) [reply test/integration/connection/test-connect-timeout.js:5]: I can change this non-routable host to a local fake server

**Comment by yuxizhe** [comment]
> @yuxizhe I've added a few comments about some of the tests because I was bumping into issues while running the test suite within a private network or without internet access. That happens due to the fact that the tests are using a hard-coded host that is not really reachable.
> 
> I was thinking about changing them but I'm still a bit skeptical about their nature, so it would be great to hear from you or @sidorares about them.

https://github.com/sidorares/node-mysql2/pull/1378  I made a PR . Use a mock server instead of hard-coded host

**Pull request merged**